File: rust/ex01-lv1/fist04/src/zero1.rs
// 구조체 기본 정의
// 여러개의 필드를 하나의 타입으로 묶어주는 역할
// 관련 있는 데이터를 함께 묶어서 사용하기에 용이함
// struct 키워드를 사용해 정의함
pub struct Person {
    // main.rs에서 사용하기 위해 구조체와 필드 앞에 pub 붙임
    pub name: String,
    pub age: u32,
    pub email: String,
}

impl Person {
    // 러스트에서 self는 자바에서 this와 유사한 역할을 함
    // 키워드가 아닌 식별자이므로 this: &self 처럼 이름을 바꿔 사용할수 있지만
    // 관례는 self를 그대로 사용하는 것임
    // 또한 self는 첫 번째 매개변수로 사용하는게 일반적 관례
    pub fn greeting(&self) -> String {
        // format!은 String 타입의 문자열을 반환함, 형식화된 문자열 생성 함수
        // 문자열을 동적으로 생성하고 반환해야 할 때 유용함.
        format!("안녕하세요, {}입니다.", self.name)
    }
}

File: python/ex01/fist03/zero1.py
# 딕셔너리: 키-값 쌍으로 데이터를 저장하는 해시 테이블 기반 구조
# 각 키는 유일하며, 해당 키에 대응하는 값을 저장함
# 매우 빠른 조회 속도, 효율적 데이터 관리 가능

# 딕셔너리 생성
person = {"name": "김박사", "age": 25, "city": "포항"}

# 값 조회
print(person["name"])  # 김박사

# 값 추가 또는 업데이트
person["job"] = "Engineer"
print(person)

# 키가 존재하는지 확인
if "age" in person:
    print("person에 age 있음")

# 값 삭제
del person["city"]
print(person)  # {'name': '김박사', 'age': 25, 'job': 'Engineer'}

# 딕셔너리 주요 메서드
# keys(), values(), items() 메서드
print(person.keys())  # dict_keys(['name', 'age', 'job'])
print(person.values())  # dict_values(['김박사', 25, 'Engineer'])
print(person.items())  # dict_items([('name', '김박사'), ('age', 25), ...

# get() 메서드: 키가 존재하지 않을 때 디폴트 값 반환
print(person.get("city", "키 없음"))  # 키 없음

# pop() 메서드: 키에 해당하는 값 반환하며 딕셔너리에서 제거
job = person.pop("job")
print(job)  # Engineer
print(person)  # {'name': '김박사', 'age': 25}

# popitem() 메서드: 마지막 키-값 쌍을 반환하며 제거
name = person.popitem()
print(name)  # ('age', 25)
print(person)  # {'name': '김박사'}

# clear() 메서드: 모든 항목을 제거
person.clear()
print(person)  # {}

# 딕셔너리 컴프리헨션
# 짧고 간결하게 딕셔너리를 생성하는 방법
squares = {x: x**2 for x in range(4)}
print(squares)  # {0: 0, 1: 1, 2: 4, 3: 9}
